File: backend_python/src/todo_api/application_service/usecase/get_all_todos_usecase.py
from abc import ABC, abstractmethod

from pydantic import BaseModel, ConfigDict

from todo_api.domain.model.todo import Todo
from todo_api.domain.repository.todo_repository import TodoRepository


class GetAllTodosUsecaseInput(BaseModel):
    pass


class GetAllTodosUsecaseOutput(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    todos: list[Todo]


class GetAllTodosUsecase(ABC):
    @abstractmethod
    def execute(self, input_dto: GetAllTodosUsecaseInput) -> GetAllTodosUsecaseOutput:
        pass


class GetAllTodosUsecaseImpl(GetAllTodosUsecase):
    def __init__(self, todo_repository: TodoRepository) -> None:
        self.todo_repository = todo_repository

    def execute(self, input_dto: GetAllTodosUsecaseInput) -> GetAllTodosUsecaseOutput:
        todos = self.todo_repository.find_all()
        return GetAllTodosUsecaseOutput(todos=todos)

File: backend_bun/src/application_service/usecase/getAllTodosUsecase.ts
import type { Todo } from "@/domain/model/todo";
import type { TodoRepository } from "@/domain/repository/todoRepository";

export type GetAllTodosResponse = {
  todos: Todo[];
};

export interface GetAllTodosUsecase {
  execute(): Promise<GetAllTodosResponse>;
}

export class GetAllTodosUsecaseImpl implements GetAllTodosUsecase {
  constructor(private todoRepository: TodoRepository) {}

  async execute(): Promise<GetAllTodosResponse> {
    const todos = await this.todoRepository.findAll();
    return { todos };
  }
}
